File: integration-tests-java/src/main/java/io/micronaut/jimmer/java/it/repository/BookRepository.java
package io.micronaut.jimmer.java.it.repository;

import io.micronaut.data.model.Pageable;
import io.micronaut.jimmer.java.it.entity.Book;
import io.micronaut.jimmer.java.it.entity.Tables;
import io.micronaut.jimmer.repository.DynamicParam;
import io.micronaut.jimmer.repository.JRepository;
import io.micronaut.jimmer.repository.annotation.Repository;
import java.math.BigDecimal;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import org.babyfish.jimmer.Page;
import org.babyfish.jimmer.sql.ast.Expression;
import org.babyfish.jimmer.sql.ast.tuple.Tuple2;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.jetbrains.annotations.Nullable;

@Repository
public interface BookRepository extends JRepository<Book, Long> {

    default Book selectBookById(long id) {
        return sql().findById(Book.class, id);
    }

    default Map<Long, List<Long>> findNewestIdsGroupByStoreId(Collection<Long> ids) {
        return Tuple2.toMultiMap(
                sql().createQuery(Tables.BOOK_TABLE)
                        .where(
                                Expression.tuple(
                                                Tables.BOOK_TABLE.name(),
                                                Tables.BOOK_TABLE.edition())
                                        .in(
                                                sql().createSubQuery(Tables.BOOK_TABLE)
                                                        .where(Tables.BOOK_TABLE.storeId().in(ids))
                                                        .groupBy(Tables.BOOK_TABLE.name())
                                                        .select(
                                                                Tables.BOOK_TABLE.name(),
                                                                Tables.BOOK_TABLE.edition().max())))
                        .select(Tables.BOOK_TABLE.storeId(), Tables.BOOK_TABLE.id())
                        .execute());
    }

    Book findByNameAndEditionAndPrice(
            String name, int edition, BigDecimal price, Fetcher<Book> fetcher);

    List<Book> findByNameLike(String name, Fetcher<Book> fetcher);

    List<Book> findByStoreId(Long storeId, Fetcher<Book> fetcher);

    Page<Book> findByNameLikeOrderByName(String name, Pageable pageable, Fetcher<Book> fetcher);

    Page<Book> findByNameLikeIgnoreCaseAndStoreNameOrderByNameAscEditionDesc(
            Pageable pageable,
            @Nullable Fetcher<Book> fetcher,
            @DynamicParam @Nullable String name,
            @Nullable String storeName);
}

File: integration-tests-kotlin/src/main/kotlin/io/micronaut/jimmer/kotlin/it/repository/BookRepository.kt
package io.micronaut.jimmer.kotlin.it.repository

import io.micronaut.data.model.Pageable
import io.micronaut.jimmer.kotlin.it.entity.Book
import io.micronaut.jimmer.repository.DynamicParam
import io.micronaut.jimmer.repository.KRepository
import io.micronaut.jimmer.repository.annotation.Repository
import org.babyfish.jimmer.Page
import org.babyfish.jimmer.sql.fetcher.Fetcher
import java.math.BigDecimal

@Repository
interface BookRepository : KRepository<Book, Long> {
    fun selectBookById(id: Long): Book? = sql.findById(Book::class, id)

    fun findByNameAndEditionAndPrice(
        name: String,
        edition: Int,
        price: BigDecimal,
        fetcher: Fetcher<Book>,
    ): Book

    fun findByNameLike(
        name: String,
        fetcher: Fetcher<Book>,
    ): List<Book>

    fun findByStoreId(
        storeId: Long,
        fetcher: Fetcher<Book>,
    ): List<Book>

    fun findByNameLikeOrderByName(
        name: String,
        pageable: Pageable,
        fetcher: Fetcher<Book>,
    ): Page<Book>

    fun findByNameLikeIgnoreCaseAndStoreNameOrderByNameAscEditionDesc(
        pageable: Pageable,
        fetcher: Fetcher<Book>,
        @DynamicParam name: String?,
        storeName: String?,
    ): Page<Book>
}
